src/validation.cpp

```diff
@@ -417,21 +418,135 @@ static bool CheckInputsFromMempoolAndCache(const CTransaction& tx, CValidationSt
+    // utxo set or in the mempool.
+    bool ConsensusScriptChecks(ATMPArgs& args, const CTransactionRef& ptx, Workspace& ws, PrecomputedTransactionData &txdata) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
+
+    // Try to add the transaction to the mempool, removing any conflicts first.
+    // Returns true if the transaction is in the mempool after any size
+    // limiting is performed, false otherwise.
+    bool Finalize(ATMPArgs& args, const CTransactionRef& ptx, Workspace& ws) EXCLUSIVE_LOCKS_REQUIRED(cs_main, pool.cs);
+
+    // Compare a package's feerate against minimum allowed.
+    bool CheckFeeRate(size_t package_size, CAmount package_fee, CValidationState& state)
+    {
+        CAmount mempoolRejectFee = pool.GetMinFee(gArgs.GetArg("-maxmempool", DEFAULT_MAX_MEMPOOL_SIZE) * 1000000).GetFee(package_size);
+        if (mempoolRejectFee > 0 && package_fee < mempoolRejectFee) {
+            return state.Invalid(ValidationInvalidReason::TX_MEMPOOL_POLICY, false, REJECT_INSUFFICIENTFEE, "mempool min fee not met", strprintf("%d < %d", package_fee, mempoolRejectFee));
+        }
+
+        if (package_fee < ::minRelayTxFee.GetFee(package_size)) {
+            return state.Invalid(ValidationInvalidReason::TX_MEMPOOL_POLICY, false, REJECT_INSUFFICIENTFEE, "min relay fee not met", strprintf("%d < %d", package_fee, ::minRelayTxFee.GetFee(package_size)));
+        }
+        return true;
+    }
+
+private:
+    CTxMemPool& pool;
+    CCoinsViewCache view;
+    CCoinsViewMemPool viewMemPool;
+    CCoinsView dummy;
+
+    // The package limits in effect at the time of invocation.
+    size_t nLimitAncestors{0};
+    size_t nLimitAncestorSize{0};
+    size_t nLimitDescendants{0};
+    size_t nLimitDescendantSize{0};
+};
+
+bool MemPoolAccept::PreChecks(ATMPArgs& args, const CTransactionRef& ptx, Workspace& ws)
 {
     const CTransaction& tx = *ptx;
-    const uint256 hash = tx.GetHash();
-    AssertLockHeld(cs_main);
-    LOCK(pool.cs); // mempool "read lock" (held through GetMainSignals().TransactionAddedToMempool())
+    const uint256& hash = ws.hash;
+
+    // Copy/alias what we need out of args
+    CValidationState &state = args.state;
+    bool* pfMissingInputs = args.pfMissingInputs;
+    const int64_t nAcceptTime = args.nAcceptTime;
+    const bool bypass_limits = args.bypass_limits;
+    const CAmount& nAbsurdFee = args.nAbsurdFee;
+    std::vector<COutPoint>& coins_to_uncache = args.coins_to_uncache;
+
+    // Alias what we need out of ws
+    std::set<uint256>& setConflicts = ws.setConflicts;
+    CTxMemPool::setEntries& allConflicting = ws.allConflicting;
+    CTxMemPool::setEntries& setAncestors = ws.setAncestors;
+    std::unique_ptr<CTxMemPoolEntry>& entry = ws.entry;
+    bool& fReplacementTransaction = ws.fReplacementTransaction;
+    CAmount& nModifiedFees = ws.nModifiedFees;
+    CAmount& nConflictingFees = ws.nConflictingFees;
+    size_t& nConflictingSize = ws.nConflictingSize;
```

Comment:
Given that the local name and the name of the member are identical, does it really make sense to alias it? Looks like you already changed every line in this function due to whitespace fixups, so might as well prefix what we need with `ws.` or `args.` in line. No strong opinion, though.
If you decide to keep the current approach, I think you can make review easier by not duplicating all the types (use `auto&`, which should keep the type and constness as is.